block-storage/src/main/scala/coop/rchain/blockstorage/casperbuffer/CasperBufferKeyValueStorage.scala

```diff
@@ -0,0 +1,107 @@
+package coop.rchain.blockstorage.casperbuffer
+
+import cats.effect.Concurrent
+import cats.syntax.all._
+import coop.rchain.catscontrib.Catscontrib._
+import coop.rchain.blockstorage.dag.codecs._
+import coop.rchain.shared.syntax._
+import coop.rchain.blockstorage.util.{
+  BlockDependencyDag,
+  DoublyLinkedDag,
+  DoublyLinkedDagOperations
+}
+import cats.instances.list._
+import coop.rchain.metrics.Metrics
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.shared.{Cell, Log}
+import coop.rchain.store.{KeyValueStoreManager, KeyValueTypedStore}
+import scodec.Codec
+
+/**
+  * @param parentsStore - persistent map {hash -> parents set}
+  * @param blockDependencyDag - in-memory dependency DAG, recreated from parentsStore on node startup
+  */
+final class CasperBufferKeyValueStorage[F[_]: Concurrent: Log] private (
+    parentsStore: KeyValueTypedStore[F, BlockHash, Set[BlockHash]],
+    blockDependencyDag: Cell[F, DoublyLinkedDag[BlockHash]]
+) extends CasperBufferStorage[F] {
+
+  override def addRelation(parent: BlockHash, child: BlockHash): F[Unit] =
+    for {
+      parents <- parentsStore.get(child).flatMap(_.getOrElse(Set.empty[BlockHash]).pure[F])
+      _       <- parentsStore.put(child, parents + parent)
+      _ <- blockDependencyDag.modify(
+            curState => DoublyLinkedDagOperations.add(curState, parent, child)
+          )
+    } yield ()
+
+  override def remove(hash: BlockHash): F[Unit] =
+    for {
+      curState                                  <- blockDependencyDag.read
+      (newState, hashesAffected, hashesRemoved) = DoublyLinkedDagOperations.remove(curState, hash)
+      _                                         <- blockDependencyDag.set(newState)
+      changes                                   = hashesAffected.map(h => (h, newState.childToParentAdjacencyList(h))).toSeq
+      deletes                                   = hashesRemoved.toSeq
+      _                                         <- parentsStore.put(changes)
+      // If node crush here - KV store end up in inconsistent state. What's the good way to handle it?
+      _ <- parentsStore.delete(deletes)
+    } yield ()
+
+  override def getParents(blockHash: BlockHash): F[Option[Set[BlockHash]]] =
+    blockDependencyDag.read.flatMap(_.childToParentAdjacencyList.get(blockHash).pure[F])
+
+  override def getChildren(blockHash: BlockHash): F[Option[Set[BlockHash]]] =
+    blockDependencyDag.read.flatMap(_.parentToChildAdjacencyList.get(blockHash).pure[F])
+
+  override def getPendants: F[Set[BlockHash]] =
+    blockDependencyDag.read.flatMap(_.dependencyFree.pure[F])
+
+  // Block is considered to be in CasperBuffer when there is a records about its parents
+  override def contains(blockHash: BlockHash): F[Boolean] =
+    parentsStore.contains(blockHash) &&^ parentsStore.get(blockHash).flatMap(_.get.nonEmpty.pure[F])
```

Comment:
You are right, better to check `blockDependencyDag` as we have it.
But should not `&&^` be lazy? so if `contains` on KV store return false, `get` is not executed?

Reply:
Yes you are right, for _false_ case execution is the same.

Reply:
@tgrospic I just figured out what you meant by `Only genesis can be without parents, for which this will return false`. 

Let me clarify something. Block first end up in buffer when it is met as a dependency. At that point it is inserted as some another block dependancy and is treated as pendant block. https://github.com/rchain/rchain/blob/6424ec9a028795840141822a725ea04996b91431/block-storage/src/main/scala/coop/rchain/blockstorage/casperbuffer/CasperBufferKeyValueStorage.scala#L63-L64
It does not have any parents - because BlockRetriever is in process of acquiring them from the network.
So there is plenty of situations when parentsStore does not have hash as a key - node just did not receive dependencies yet.